### PR TITLE
feat(manifest): Add target and cache from build options

### LINKED ISSUE #1331: Support multi-stage build targets
I may be going about this wrong, but I have a container which needs asset dependencies installed and compiled for a production build but not local/development builds. Specifically, I need to install npm dependencies and run gulp.

I couldn't think of a way to accomplish this, other than to create a multi-stage build in the Dockerfile and attempt to reference the specific stage but I don't see a way to do that with the init command.

If I have missed something or am attempting some grave sin by attempting this, I welcome any feedback.

Cheers!

### PR DESCRIPTION
**Pull request by mRoca**
This PR allows to specify the `--target` and `--cache-from` docker build options with a [docker-compose](https://docs.docker.com/compose/compose-file/#cache_from) like syntaxe

```yaml
image:
  build:
    dockerfile: path/to/dockerfile
    context: context/dir
    target: build-stage
    cache_from:
      - image:tag
    args:
      key: value
```

The `cache_from` option allows to pull a previously built image and use it as cache layers, and the `target` option allows... well, to use a multi-stages Dockerfile.

Resolves #1331

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by efekarakus** [comment]
This is awesome thanks @mRoca ! it also resolves https://github.com/aws/copilot-cli/issues/1331

File: internal/pkg/docker/docker_test.go [hunk 1]
```diff
@@ -103,6 +105,20 @@ func TestBuild(t *testing.T) {
 					"mockPath/to", "-f", "mockPath/to/mockDockerfile"}).Return(nil)
 			},
 		},
+		"success with options": {
```
Comment (efekarakus) [inline internal/pkg/docker/docker_test.go]: nit:
```suggestion
		"runs with cache_from and target fields": {
```

File: internal/pkg/manifest/workload.go [hunk 2]
```diff
@@ -111,6 +107,17 @@ func (i *Image) args() map[string]string {
 	return i.Build.BuildArgs.Args
 }
 
+// target returns the build target stage if it exists, otherwise nil.
+func (i *Image) target() *string {
+	return i.Build.BuildArgs.Target
+}
+
+// cacheFrom returns the cache from build section, if it exists.
+// Otherwise it returns an empty array.
```
Comment (efekarakus) [inline internal/pkg/manifest/workload.go]:
```suggestion
// Otherwise it returns nil.
```

File: internal/pkg/manifest/workload_test.go [hunk 3]
```diff
@@ -56,6 +56,22 @@ func TestBuildArgs_UnmarshalYAML(t *testing.T) {
 				},
 			},
 		},
+		"Dockerfile with cache from and target build opts": {
```
Comment (efekarakus) [inline internal/pkg/manifest/workload_test.go:59]: Great tests thank you so much!

File: site/content/docs/manifest/lb-web-service.md [hunk 4]
```diff
@@ -71,10 +71,13 @@ image:
...
     dockerfile: path/to/dockerfile
     context: context/dir
+    target: build-stage
+    cache_from:
+      - image:tag
     args:
       key: value
 ```
-In this case, copilot will use the context directory you specified and convert the key-value pairs under args to --build-arg overrides. The equivalent docker build call will be: `$ docker build --file path/to/dockerfile --build-arg key=value context/dir`.
+In this case, copilot will use the context directory you specified and convert the key-value pairs under args to --build-arg overrides. The equivalent docker build call will be: `$ docker build --file path/to/dockerfile --target build-stage --cache-from image:tag --build-arg key=value context/dir`.
```
Comment (efekarakus) [inline site/content/docs/manifest/lb-web-service.md:80]: Would you mind adding this blurb to the new `manifest/scheduled-job.md` as well?

File: templates/cicd/buildspec.yml [hunk 5]
```diff
@@ -86,6 +88,14 @@ phases:
...
             done
           fi
+          if [ -n "$build_target" ]; then
+            build_args="$build_args--target $build_target "
+          fi
+          if [ -n "$build_cache_from" ]; then
+            for arg in $(echo $build_cache_from | jq -r '.[]'); do
+              build_args="$build_args--cache-from $arg "
+            done
+          fi
```
Comment (efekarakus) [inline templates/cicd/buildspec.yml:98]: amazing! thank you so much for diving into this and figuring out how to parse an array

File: templates/workloads/jobs/scheduled-job/manifest.yml [hunk 6]
```diff
@@ -10,7 +10,7 @@ type: {{.Type}}
 
 image:
 {{- if .ImageConfig.Build.BuildArgs.Dockerfile}}
-  # Docker build arguments. You can specify additional overrides here. Supported: dockerfile, context, args.
+  # Docker build arguments. You can specify additional overrides here. Supported: dockerfile, context, args, target, cache_from.
```
Comment (efekarakus) [inline templates/workloads/jobs/scheduled-job/manifest.yml]: What do you think of just linking to the website here instead (same for the other manifest files):
```suggestion
  # Docker build arguments. 
  # For additional overrides: https://aws.github.io/copilot-cli/docs/manifest/scheduled-job/#image-build
```

**Review by efekarakus — COMMENTED** [review]
Amazing! Only small text recommendations but otherwise the code looks great 👌

File: templates/workloads/jobs/scheduled-job/manifest.yml (hunk 6 above)
Comment (mRoca) [reply templates/workloads/jobs/scheduled-job/manifest.yml]: Done, good idea :+1:

File: internal/pkg/manifest/workload.go (hunk 2 above)
Comment (mRoca) [reply internal/pkg/manifest/workload.go]: Good catch

**Review by efekarakus — APPROVED** [review]
Awesome, thank you!

**Review by bvtujo — APPROVED** [review]
This looks awesome, thank you so much for the contribution!

**Pull request merged**